src/_posts/2020-05-19-promises-and-lies.md

```diff
@@ -0,0 +1,151 @@
+tags:
+    - javascript
+    - node
+    - reliability
+abstract: Adding timeouts to promises in Javascript. Introducing lie.js 
+alias: [/2020/05/19/promises-and-lies]
+---
+
+## Promises in Javascript
+
+Promises were added to Javascript as a standard means to avoid nesting callback functions.
+Before promises each function could have it's own functions that would be called once the job is completed. They are mostly used for asynchronous events. They are not a means of lazy evaluation or for process control. Functions that have an async result typically now return a Promise.
+
+Promises have the following format:
+
+```javascript
+var promiseObj = new Promise(executor);
+```
+
+where executor is a function that looks like:
+
+```javascript
+function(resolutionFunc, rejectionFunc){
+    // typically, some asynchronous operation.
+}
+``` 
+
+Promises are used like this:
+
+```javascript
+promiseObj.then(console.log('It worked'))
+          .catch(console.log('It failed'))
+          .finally( console.log('It is over') )
+```
+
+You would typically do more in each block but this gives the idea. It is possible to return a promise from a promise so that they can be chained. Note that the catch at the end of a chain will catch whichever reject happens first.
+
+Once a promise has resolved to a value, then it will always resolve to the same value.
+You can therefore call then on a given promise multiple times and it will always return the same result. I'll come back to this concept later.
+
+Promises also have no means of being cancelled. The only output of them is through actions of the resolve or reject functions. This means that it is possible to cache a promise - just note that the cached promise could have already failed!
+
+## The Promise class
+
+The javascript promise class has the following static functions:
+
+```Javascript
+Promise.all( ) 
+```
+
+This will turn an iteratable of Promises into a Promise that will either return an iteratable of the success values or the first failure condition.
+
+```Javascript
+Promise.allSettled()
+```
+
+This will take an iteratable of Promises and returns a Promise that will return an iteratable of the results (either resolved or rejected).
+
+```Javascript
+Promise.race()
+```
+
+This will take an iteratable of Promises and returns a Promise that will return the result of the first resolved or rejected Promise.
+
+```Javascript
+Promise.reject()
+```
+
+This returns a Promise that has been rejected with the given reason.  
+
+```Javascript
+Promise.resolve()
+```
+
+This will either return the result of a supplied promise or will return a Promise that has resolved the supplied data.
+
+## What is missing?
+
+The standard library has no built-in concept of timeout. This means that each usage must consider (or more likely ignore) this important concept. Without a controlled timeout you have no means of ensuring that a given ```then()``` will ever return.
```

Comment:
Does this need three backticks? No big deal but if it does I'll have to adjust to that. 

```suggestion
The standard library has no built-in concept of timeout. This means that each usage must consider (or more likely ignore) this important concept. Without a controlled timeout you have no means of ensuring that a given `then()` will ever return.
```